scenario.py

```diff
@@ -189,10 +190,215 @@ def __init__(self, params, experiment_path):
+        for label in labels:
+            idx_in_full_trainset = np.where(y_train == label)
+            x_train_for_cluster[label] = x_train[idx_in_full_trainset]
+            y_train_for_cluster[label] = y_train[idx_in_full_trainset]
+            count_available_per_cluster[label] = len(y_train_for_cluster[label])
+
+        # Sort partners per nb of different clusters in descending order
+        partners_with_shared_clusters = sorted(
+            partners_with_shared_clusters,
+            key=operator.attrgetter("cluster_count"),
+            reverse=True,
+        )
+        partners_with_specific_clusters = sorted(
+            partners_with_specific_clusters,
+            key=operator.attrgetter("cluster_count"),
+            reverse=True,
+        )
+
+        # Initialize dict for storing which partners pick from which clusters
+        dict_specific, dict_shared = {}, {}
+
+        # First, partners getting data samples from clusters they don't share
+        position = 0
+        for p in partners_with_specific_clusters:
+            dict_specific[p.id] = {"clusters_list": labels[position:position + p.cluster_count]}
+            position += p.cluster_count
+
+        # Second, partners getting data samples from shared clusters
+        shared_clusters = labels[position:position + max_shared_clusters]
+        for p in partners_with_shared_clusters:
+            dict_shared[p.id] = {"clusters_list": random.sample(shared_clusters, k=p.cluster_count)}
+
+        # print(shared_clusters)  # DEBUG
+        # print(dict_specific)  # DEBUG
+        # print(dict_shared)  # DEBUG
+
+        # Compute the data amount limiting factor
+        resize_factor = 1
+
+        # For partners getting data samples from clusters they don't share
+        # ... compare the nb of available samples vs. the nb of samples initially configured
+        for p in partners_with_specific_clusters:
+            count_available = sum([count_available_per_cluster[cl] for cl in dict_specific[p.id]["clusters_list"]])
+            count_initialconfig = int(amounts_per_partner[p.id] * len(y_train))
+            dict_specific[p.id]["nb_available_samples"] = count_available
+            dict_specific[p.id]["nb_samples_configured"] = count_initialconfig
+            ratio = count_available / count_initialconfig
+            resize_factor = min(resize_factor, ratio)
+
+        # print(dict_specific)  # DEBUG
+        # print("resize_factor after specific: ", resize_factor)  # DEBUG
+
+        # For each partner getting data samples from shared clusters:
+        # ... compute the nb of samples initially configured and resize it,
+        # ... then sum per cluster how many samples are needed
+        nb_samples_needed_per_cluster = dict.fromkeys(shared_clusters, 0)
+        # print(nb_samples_needed_per_cluster)  # DEBUG
+        for p in partners_with_shared_clusters:
+            count_initialconfig_resized = int(amounts_per_partner[p.id] * len(y_train) * resize_factor)
```

Comment:
count_initialconfig_resized ??

Reply:
Ok :-) Renamed to `initial_amount_resized`